src/main/java/com/google/devtools/build/lib/exec/RunfilesTreeUpdater.java

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.exec;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.devtools.build.lib.actions.Artifact;
+import com.google.devtools.build.lib.actions.ArtifactPathResolver;
+import com.google.devtools.build.lib.actions.ExecException;
+import com.google.devtools.build.lib.actions.RunfilesSupplier;
+import com.google.devtools.build.lib.actions.UserExecException;
+import com.google.devtools.build.lib.analysis.RunfilesSupport;
+import com.google.devtools.build.lib.profiler.Profiler;
+import com.google.devtools.build.lib.profiler.ProfilerTask;
+import com.google.devtools.build.lib.util.io.OutErr;
+import com.google.devtools.build.lib.vfs.FileStatus;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utility used in local execution to create a runfiles tree if {@code --nobuild_runfile_links}
+ * has been specified.
+ */
+@ThreadSafe
+public class RunfilesTreeUpdater {
+
+  /**
+   * Poor man's reference counted object pool.
+   *
+   * <p>Maintains a mapping of runfiles directory to monitor. The monitor is an AtomicInteger in
+   * order to keep track how often the monitor has been acquired. If the count drops to 0 the
+   * mapping will be removed.
+   */
+  private final ConcurrentMap<PathFragment, AtomicInteger> monitorsWithRefcnt
+      = new ConcurrentHashMap<>();
+  private final boolean enableRunfiles;
+
+
+  public RunfilesTreeUpdater(boolean enableRunfiles) {
+    this.enableRunfiles = enableRunfiles;
+  }
+
+  private static void updateRunfilesTree(Path execRoot, PathFragment runfilesDir,
+      BinTools binTools, ImmutableMap<String, String> env, OutErr outErr, boolean enableRunfiles)
+      throws IOException, ExecException {
+    Path runfilesDirPath = execRoot.getRelative(runfilesDir);
+    Path inputManifest = RunfilesSupport.inputManifestPath(runfilesDirPath);
+    if (!inputManifest.exists()) {
+      // TODO(buchgr): Add logic to create a runfiles tree if --nobuild_runfile_manifests is set
+      throw new UserExecException("Can't build local runfiles tree because "
+          + "--nobuild_runfile_manifests is set.");
+    }
+
+    Path outputManifest = RunfilesSupport.outputManifestPath(runfilesDirPath);
+    try {
+      // Avoid rebuilding the runfiles directory if the manifest in it matches the input manifest,
+      // implying the symlinks exist and are already up to date. If the output manifest is a
+      // symbolic link, it is likely a symbolic link to the input manifest, so we cannot trust it as
```

Comment:
Is it? I thought that we create it anew so as to avoid this exact issue, don't we?

Reply:
I copied this comment from the old code. I also could not find a place where the output manifest is a symlink to the input manifest. It might be time to retire the comment and remove the check?